assets/man/man3/wx_utils.3
.TH "wx_utils" 3 "ESP-IDF Components by K0I05" \" -*- nroff -*-
.ad l
.nh
.SH NAME
wx_utils \- Wx_utils
.SH SYNOPSIS
.br
.PP
.SS "Data Structures"

.in +1c
.ti -1c
.RI "struct \fBwx_scalar_range_t\fP"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "esp_err_t \fBwx_set_temperature_range\fP (const float maximum, const float minimum)"
.br
.RI "Sets the global weather utilities maximum and minimum temperature range (\fRwx_temperature_range\fP) in degrees Celsius\&. "
.ti -1c
.RI "esp_err_t \fBwx_set_humidity_range\fP (const float maximum, const float minimum)"
.br
.RI "Sets the global weather utilities humidity maximum and minimum range (\fRwx_humidity_range\fP) in percent\&. "
.ti -1c
.RI "esp_err_t \fBwx_td_range\fP (const float temperature, const float humidity, float *const dewpoint)"
.br
.RI "Calculates dewpoint temperature from air temperature and relative humidity with range validation\&. The default temperature range is from -40 to 125 degrees Celsius and default humidity range is from 0 to 100 percent\&. 
.br
 The default ranges can be adjusted through the set range functions\&. "
.ti -1c
.RI "const double \fBwx_c_to_k\fP (const double t)"
.br
.RI "Converts degrees celsius to kelvin\&. "
.ti -1c
.RI "const double \fBwx_k_to_c\fP (const double t)"
.br
.RI "Converts kelvin to degrees celsius\&. "
.ti -1c
.RI "const double \fBwx_v\fP (const double t)"
.br
.RI "Calculates v at temperature\&. "
.ti -1c
.RI "const double \fBwx_pwi\fP (const double t)"
.br
.RI "Calculates aqueous vapor pressure of ice at temperature\&. "
.ti -1c
.RI "const double \fBwx_pws\fP (const double t)"
.br
.RI "Calculates aqueous vapor pressure of water at temperature\&. "
.ti -1c
.RI "const double \fBwx_pressure_at_sea_level\fP (const double pa, const double l, const double ta)"
.br
.RI "Calculates the reduced air pressure QFF (pressure at sea level) without QFE\&. "
.ti -1c
.RI "const double \fBwx_qfe\fP (const double pa, const double l, const double ta)"
.br
.RI "Calculates the reduced air pressure QFE (pressure at a certain level)\&. "
.ti -1c
.RI "const double \fBwx_qff\fP (const double qfe, const double l, const double ta)"
.br
.RI "Calculates the reduced air pressure QFF (pressure at sea level)\&. "
.ti -1c
.RI "const double \fBwx_qnh\fP (const double qfe, const double h, const double a)"
.br
.RI "Calculates the reduced air pressure QNH (pressure at sea level according to ICAO standard atmospheric)\&. "
.ti -1c
.RI "const double \fBwx_td\fP (const double ta, const double hr)"
.br
.RI "Calculates dewpoint temperature from air temperature and relative humidity\&. "
.ti -1c
.RI "const double \fBwx_tw\fP (const double ta, const double hr, const double td, const double pa)"
.br
.RI "Calculates wetbulb temperature from air temperature, relative humidity, and dewpoint temperature\&. "
.in -1c
.SH "Detailed Description"
.PP 
ESP-IDF weather utilities

.PP
Ported from esp-open-rtos

.PP
Copyright (c) 2024 Eric Gionet (gionet.c.eric@gmail.com)

.PP
MIT Licensed as described in the file LICENSE 
.SH "Function Documentation"
.PP 
.SS "const double wx_c_to_k (const double t)"

.PP
Converts degrees celsius to kelvin\&. 
.PP
\fBParameters\fP
.RS 4
\fIt\fP Temperature in degrees celsius\&. 
.RE
.PP
\fBReturns\fP
.RS 4
double Temperature in kelvin\&. 
.RE
.PP

.SS "const double wx_k_to_c (const double t)"

.PP
Converts kelvin to degrees celsius\&. 
.PP
\fBParameters\fP
.RS 4
\fIt\fP Temperature in kelvin\&. 
.RE
.PP
\fBReturns\fP
.RS 4
double Temperature in degrees celsius\&. 
.RE
.PP

.SS "const double wx_pressure_at_sea_level (const double pa, const double l, const double ta)"

.PP
Calculates the reduced air pressure QFF (pressure at sea level) without QFE\&. 
.PP
\fBParameters\fP
.RS 4
\fIpa\fP Air pressure at this altitude in hecto-pascal\&. 
.br
\fIl\fP Reduction level at this altitude in meters\&. 
.br
\fIta\fP Air temperature at this altitude in degrees celsius\&. 
.RE
.PP
\fBReturns\fP
.RS 4
float Reduced air pressure QFF in hecto-pascal\&. 
.RE
.PP

.SS "const double wx_pwi (const double t)"

.PP
Calculates aqueous vapor pressure of ice at temperature\&. 
.PP
\fBParameters\fP
.RS 4
\fIt\fP Temperature in kelvin\&. 
.RE
.PP
\fBReturns\fP
.RS 4
double Aqueous vapor pressure of ice 
.RE
.PP

.SS "const double wx_pws (const double t)"

.PP
Calculates aqueous vapor pressure of water at temperature\&. 
.PP
\fBParameters\fP
.RS 4
\fIt\fP Temperature in kelvin\&. 
.RE
.PP
\fBReturns\fP
.RS 4
double Aqueous vapor pressure of water\&. 
.RE
.PP

.SS "const double wx_qfe (const double pa, const double l, const double ta)"

.PP
Calculates the reduced air pressure QFE (pressure at a certain level)\&. 
.PP
\fBNote\fP
.RS 4
The reduction level for QFE processing is the elevation difference of the pressure sensor and the QFE level into which the pressure will be reduced\&. If the pressure sensor is above the QFE level, the reduction level to get the QFE level pressure is a positive value\&. If the sensor is below the QFE level, the reduction level is a negative value\&.
.RE
.PP
\fBParameters\fP
.RS 4
\fIpa\fP Air pressure at this altitude in hecto-pascal\&. 
.br
\fIl\fP Reduction level in meters\&. 
.br
\fIta\fP Air temperature at this altitude in degrees celsius\&. 
.RE
.PP
\fBReturns\fP
.RS 4
float Reduced air pressure QFE in hecto-pascal\&. 
.RE
.PP

.SS "const double wx_qff (const double qfe, const double l, const double ta)"

.PP
Calculates the reduced air pressure QFF (pressure at sea level)\&. 
.PP
\fBNote\fP
.RS 4
The reduction level is the elevation difference of the station altitude and the mean sea level\&.
.RE
.PP
\fBParameters\fP
.RS 4
\fIqfe\fP Field elevation aire pressure in hecto-pascal\&. 
.br
\fIl\fP Reduction level at this altitude in meters\&. 
.br
\fIta\fP Air temperature at this altitude in degrees celsius\&. 
.RE
.PP
\fBReturns\fP
.RS 4
float Reduced air pressure QFF in hecto-pascal\&. 
.RE
.PP

.SS "const double wx_qnh (const double qfe, const double h, const double a)"

.PP
Calculates the reduced air pressure QNH (pressure at sea level according to ICAO standard atmospheric)\&. 
.PP
\fBParameters\fP
.RS 4
\fIqfe\fP Staion level air pressure in hecto-pascal\&. 
.br
\fIh\fP Elevation of pressure QFE in International Standard Atmosphere (ISA)\&. 
.br
\fIa\fP Station altitude in meters\&. 
.RE
.PP
\fBReturns\fP
.RS 4
float Reduced air pressure QNH in hecto-pascal\&. 
.RE
.PP

.SS "esp_err_t wx_set_humidity_range (const float maximum, const float minimum)"

.PP
Sets the global weather utilities humidity maximum and minimum range (\fRwx_humidity_range\fP) in percent\&. 
.PP
\fBParameters\fP
.RS 4
\fImaximum\fP Maximum humidity range in percent\&. 
.br
\fIminimum\fP Minimum humidity range in percent\&. 
.RE
.PP
\fBReturns\fP
.RS 4
esp_err_t ESP_OK on success\&. 
.RE
.PP

.SS "esp_err_t wx_set_temperature_range (const float maximum, const float minimum)"

.PP
Sets the global weather utilities maximum and minimum temperature range (\fRwx_temperature_range\fP) in degrees Celsius\&. 
.PP
\fBParameters\fP
.RS 4
\fImaximum\fP Maximum temperature range in degrees Celsius\&. 
.br
\fIminimum\fP Minimum temperature range in degrees Celsius\&. 
.RE
.PP
\fBReturns\fP
.RS 4
esp_err_t ESP_OK on success\&. 
.RE
.PP

.SS "const double wx_td (const double ta, const double hr)"

.PP
Calculates dewpoint temperature from air temperature and relative humidity\&. 
.PP
\fBParameters\fP
.RS 4
\fIta\fP Air temperature in degrees celsius\&. 
.br
\fIhr\fP Relative humidity in percent\&. 
.RE
.PP
\fBReturns\fP
.RS 4
double Dewpoint temperature in degrees celsius\&. 
.RE
.PP

.SS "esp_err_t wx_td_range (const float temperature, const float humidity, float *const dewpoint)"

.PP
Calculates dewpoint temperature from air temperature and relative humidity with range validation\&. The default temperature range is from -40 to 125 degrees Celsius and default humidity range is from 0 to 100 percent\&. 
.br
 The default ranges can be adjusted through the set range functions\&. 
.PP
\fBParameters\fP
.RS 4
\fItemperature\fP Air temperature in degrees Celsius\&. 
.br
\fIhumidity\fP Relative humidity in percent\&. 
.br
\fIdewpoint\fP Calculated dewpoint temperature in degrees Celsius\&. 
.RE
.PP
\fBReturns\fP
.RS 4
esp_err_t ESP_OK on success\&. 
.RE
.PP

.SS "const double wx_tw (const double ta, const double hr, const double td, const double pa)"

.PP
Calculates wetbulb temperature from air temperature, relative humidity, and dewpoint temperature\&. 
.PP
\fBParameters\fP
.RS 4
\fIta\fP Air temperature in degrees celsius\&. 
.br
\fIhr\fP Relative humidity in percent\&. 
.br
\fItd\fP Dewpoint temperature in degrees celsius\&. 
.br
\fIpa\fP Air pressure in hecto-pascal\&. 
.RE
.PP
\fBReturns\fP
.RS 4
double Wetbulb temperature in degrees celsius\&. 
.RE
.PP

.SS "const double wx_v (const double t)"

.PP
Calculates v at temperature\&. 
.PP
\fBParameters\fP
.RS 4
\fIt\fP Temperature in kelvin\&. 
.RE
.PP
\fBReturns\fP
.RS 4
double v at temperature\&. 
.RE
.PP

.SH "Author"
.PP 
Generated automatically by Doxygen for ESP-IDF Components by K0I05 from the source code\&.


components/utilities/esp_wx_utils/wx_utils.c
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2024 Eric Gionet (gionet.c.eric@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * @file wx_utils.c
 *
 * ESP-IDF weather utilities (utils)
 *
 * References
 * https://docs.vaisala.com/r/M212417EN-H/en-US/GUID-EBE2B115-3C98-4C8C-9F2D-A2FF1EFECFCC
 * https://earthscience.stackexchange.com/questions/16366/weather-forecast-based-on-pressure-temperature-and-humidity-only-for-implement
 * https://web.archive.org/web/20110610213848/http://www.meteormetrics.com/zambretti.htm
 *
 * Copyright (c) 2024 Eric Gionet (gionet.c.eric@gmail.com)
 *
 * MIT Licensed as described in the file LICENSE
 */

#include "wx_utils.h"
#include <string.h>
#include <stdio.h>
#include <esp_log.h>
#include <esp_check.h>
#include <freertos/FreeRTOS.h>
#include <freertos/task.h>

/*
 * weather utilities definitions
*/

#define WX_TEMPERATURE_MAX       (float)(125.0)  //!< maximum temperature range
#define WX_TEMPERATURE_MIN       (float)(-40.0)  //!< minimum temperature range
#define WX_HUMIDITY_MAX          (float)(100.0)  //!< maximum humidity range
#define WX_HUMIDITY_MIN          (float)(0.0)    //!< minimum humidity range


/*
 * macro definitions
*/

#define ESP_ARG_CHECK(VAL) do { if (!(VAL)) return ESP_ERR_INVALID_ARG; } while (0)

/*
* weather utilities static global declarations
*/

/**
 * @brief Weather utilities temperature maximum and minimum ranges in degrees Celsius (default maximum 125 and minimum -40).
 */
static inline wx_scalar_range_t wx_temperature_range = { .maximum = WX_TEMPERATURE_MAX, .minimum = WX_TEMPERATURE_MIN };

/**
 * @brief Weather utilities humidity maximum and minimum ranges in percent (default maximum 100 and minimum 0).
 */
static inline wx_scalar_range_t wx_humidity_range = { .maximum = WX_HUMIDITY_MAX, .minimum = WX_HUMIDITY_MIN };


/*
* static constant declarations
*/

static const char *TAG = "wx_utils";


/*
* functions and subroutines
*/

esp_err_t wx_set_temperature_range(const float maximum, const float minimum) {
    /* validate max and min arguments */
    if(minimum > maximum || maximum < minimum) {
        ESP_RETURN_ON_FALSE( false, ESP_ERR_INVALID_ARG, TAG, "minimum temperature must be smaller than the maximum temperature, set temperature range failed");
    }

    /* set max and min range */
    wx_temperature_range = (wx_scalar_range_t){ .maximum = maximum, .minimum = minimum };

    return ESP_OK;
}

esp_err_t wx_set_humidity_range(const float maximum, const float minimum) {
    /* validate max and min arguments */
    if(minimum > maximum || maximum < minimum) {
        ESP_RETURN_ON_FALSE( false, ESP_ERR_INVALID_ARG, TAG, "minimum humidity must be smaller than the maximum humidity, set humidity range failed");
    }

    /* set max and min ranges */
    wx_humidity_range = (wx_scalar_range_t){ .maximum = maximum, .minimum = minimum };

    return ESP_OK;
}

esp_err_t wx_td_range(const float temperature, const float humidity, float *const dewpoint) {
    /* validate arguments */
    ESP_ARG_CHECK( dewpoint );

    // validate range of temperature parameter
    if(temperature > wx_temperature_range.maximum || temperature < wx_temperature_range.minimum) {
        ESP_RETURN_ON_FALSE( false, ESP_ERR_INVALID_ARG, TAG, "temperature is out of range, calculate dewpoint failed");
    }

    // validate range of humidity parameter
    if(humidity > wx_humidity_range.maximum || humidity < wx_humidity_range.minimum) {
        ESP_RETURN_ON_FALSE( false, ESP_ERR_INVALID_ARG, TAG, "humidity is out of range, calculate dewpoint failed");
    }

    // calculate dew-point temperature
    //double H = (log10(humidity)-2)/0.4343 + (17.62*temperature)/(243.12+temperature);
    //*dewpoint = 243.12*H/(17.62-H);

    *dewpoint = wx_td(temperature, humidity);
    
    return ESP_OK;
}




const double wx_c_to_k(const double t) {
    return t + 273.15;
}

const double wx_k_to_c(const double t) {
    return t - 273.15;
}

const double wx_v(const double t) {
    const double c0 = 0.4931358;
    const double c1 = -0.0046094296;
    const double c2 = 0.000013746454;
    const double c3 = -0.000000012743214;

    return t - c0 - c1 * t - c2 * pow(t, 2) - c3 * pow(t, 3);
}

const double wx_pwi(const double t) {
    const double a0 = -5674.5359;
    const double a1 = 6.3925247;
    const double a2 = -9.677843E-03;
    const double a3 = 0.00000062215701;
    const double a4 = 2.0747825E-09;
    const double a5 = -9.484024E-13;
    const double a6 = 4.1635019;

    return exp( (a0 / t) + a1 + (a2 * t) + (a3 * pow(t, 2)) + (a4 * pow(t, 3)) + (a5 * pow(t, 4)) + (a6 * log(t)) ) * 0.01;
}

const double wx_pws(const double t) {
    const double b_1 = -5800.2206;
    const double b0 = 1.3914993;
    const double b1 = -0.048640239;
    const double b2 = 0.000041764768;
    const double b3 = -0.000000014452093;
    const double b4 = 6.5459673;
    const double v = wx_v(t);

    return exp( (b_1 / v) + b0 + (b1 * v) + (b2 * pow(v, 2)) + (b3 * pow(v, 3)) + (b4 * log(v)) ) * 0.01;
}


const double wx_pressure_at_sea_level(const double ph, const double a, const double t) {
    // see https://keisan.casio.com/exec/system/1224575267

    // precalculate altitude, corrected for the temperature lapse rate
    const double hl = a * 0.0065;

    return ph / pow(1.0-hl/(t+hl+273.15),5.257);
}

const double wx_td(const double ta, const double hr) {
    const double a1 = 243.12;
    const double a2 = 17.62;
    const double H = log( hr * exp( (a2 * ta) / (a1 + ta) ) / 100 );

    return (a1 * H) / (a2 - H);
}

static inline const double wx_phit(const double ta, const double tw, const double pa) {
    const double pws_tw = wx_pws(wx_c_to_k(tw));

    return pws_tw - 0.000662 * pa * (ta - tw);
}

static inline const double wx_pder(const double ta, const double tw, const double phit) {
    return NAN;
}

const double wx_tw(const double ta, const double hr, const double td, const double pa) {
    const double pws_td = wx_pws(wx_c_to_k(td));

    double twet_init = 243.12 * ( log(pws_td / 6.112) / (17.62 - log(pws_td / 6.112)) );
    return NAN;
}